### ISSUE TITLE
VisualStudioWorkspace could use more threading cleanup

### ISSUE THREAD

**Issue by jasonmalinowski** [dotnet/roslyn#16940]
Although #14354 made VisualStudioWorkspace better in terms of threading, there are still a bunch of icky bits:

1. My refactoring left a ProjectTracker property that was accessible for callers to avoid breaking binary compatibility. This should be called on the UI thread but CodeLens touched it on background threads. This was fine since things had already been initialized but is getting somewhat lucky.
2. Assertions that we're on the UI thread are still icky since we have our existing historical issues with such assertions.
3. Assertions for what is and isn't callable on different threads could still be more clear.

This is somewhat an open-ended bug, but a pass to clean up the issues above would make it a lot more sane.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski is there still work to be don ehere?

**Comment by jasonmalinowski** [comment]
Nope, this has all been cleaned up. (Although somewhat recently, so the bug wasn't entirely stale.)
